### ISSUE TITLE
Epic: Historacle pricing

### ISSUE THREAD

**Issue by adamewozniak** [umee-network/umee#1521]
## Summary

As an addition to the oracle module, we need to:

1. Stamp prices for every assets every X amount of blocks
2. Provide a keeper function & query that responds with the median of the last 30 days of prices
3. Provide a keeper function that takes a denomination and a price, and will return whether or not that price is within the standard deviation of the median of the last 30 days of prices

This should serve as an epic for some smaller tasks:

- [x] https://github.com/umee-network/umee/pull/1525
- [x] #1540
- [x] #1598 
- [x] #1571
- [x] #1541
- [x] #1542
- [x] #1543
- [x] #1545
- [x] #1544
- [x] #1619 
- [x] #1570
- [x] #1621 
- [ ] #1655 
---

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by toteki** [comment]
I can make this into a submodule of leverage, so the code stays fairly isolated, until we figure out what to do with it. It'll just
- Store prices every epoch
- Calculate and store median and SD each epoch, per asset
- Get stored values using a keeper method + query

That way leverage can, for every `MsgBorrow` and possibly `MsgDecollateralize` and `MsgWithdraw`, execute some custom logic.

**Comment by adamewozniak** [comment]
> I can make this into a submodule of leverage, so the code stays fairly isolated, until we figure out what to do with it. It'll just
> 
> * Store prices every `epoch`, keeping the last `num_epochs`
> * Calculate and store median and SD of stored prices each epoch, per asset
> * Get stored values using a keeper method + query
> 
> That way leverage can, for every `MsgBorrow` and possibly `MsgDecollateralize` and `MsgWithdraw`, execute some custom logic.
> 
> (Only difference from issue desc. is the storing of calculated values since they should only change on epoch)

Ehh now that I think about it, it might be cleaner to do this within the oracle module or as an x/historacle module. Separation of concerns 🤷 @toteki

**Comment by robert-zaremba** [comment]
related to: https://twitter.com/avi_eisen/status/1582368460645945345?s=46&t=OhTJAQXLelh-16GbUngmjw

**Comment by adamewozniak** [comment]
@toteki I'm going to make a couple keeper functions for you to call on the oracle side (thinking 30 days now)

**Comment by toteki** [comment]
Keeping it in oracle module works, yeah. If the historical price can be kept as a single function (like `TokenPrice`) then I can build whatever helpers are needed on the leverage side to translate like I did with the various `TokenValue` functions.

**Comment by adamewozniak** [comment]
> Keeping it in oracle module works, yeah. If the historical price can be kept as a single function (like `TokenPrice`) then I can build whatever helpers are needed on the leverage side to translate like I did with the various `TokenValue` functions.

Thinking about just doing a separate module that has hooks into the oracle module to keep the oracle module smaller & migrations cleaner

**Comment by adamewozniak** [comment]
@toteki I'm turning this into an epic for @zarazan / @rbajollari to work on, I'll add another task for wiring this to the leverage module

**Comment by robert-zaremba** [comment]
Spec should be added / created before going into implementation.

**Comment by adamewozniak** [comment]
@robert-zaremba I'll add a spec for this today / tomorrow, have just been swamped with PFv2

**Comment by adamewozniak** [comment]
closing out historacle pricing, this came out last year
